docs/tools.md

```diff
@@ -0,0 +1,101 @@
+# Tools
+
+## Introduction
+
+Tools in the context of an agent refer to additional functionalities or capabilities integrated with the agent to perform specific tasks beyond text processing.
+
+These tools extend the agent's abilities, allowing it to interact with external systems, access information, and execute actions.
+
+## Development
+
+### Writing a new tool
+
+To create a tool, the `Tool` or `DynamicTool` class must be implemented from [tools base](../src/tools/base.ts). When starting to write tools, it is recommended to implement the `Tool` class. The `DynamicTool` class is an extension of the `Tool` class that allows for dynamic forms of input. Refer to the list of [examples](#examples) for some simple examples or any of the built-in tools in order to guide the creation of new tools.
+
+Tools MUST do the following:
+
+- Implement the `Tool` class:
+
+  ```typescript
+  export class MyNewTool extends Tool<MyNewToolOutput, ToolOptions, ToolRunOptions> {
+    // tool implementation
+  }
+  ```
+
+- Be given a unique name:
+
+  ```typescript
+  name = "MyNewTool";
+  ```
+
+- Provide a natural language description of what the tool does:
+
+  ❗Important: this description is used by the agent to determine when the tool should be used. It's probably the most important aspect of your tool and you should experiment with different natural language descriptions to ensure the tool is used in the correct circumstances.
+
+  ```typescript
+  description = "Takes X action when given Y input resulting in Z output";
+  ```
+
+- Declare an input schema:
+
+  This is used to define the format of the input to your tool. The agent will formalise the natural language input(s) it has received and structure them into the fields described in the tool's input. The input schema is specified using [Zod](https://github.com/colinhacks/zod).
```

Comment:
- You can also use `JSONSchema` (but `zod` is recommended). 
- Mention that `inputSchema` must always be an object and zod effects are not supported (z.object().transform(...)`). (https://github.com/i-am-bee/bee-agent-framework/blob/main/src/internals/helpers/schema.ts#L30-L41)
- `inputSchema` must be a function (either `async` or `sync`, but it must be a function due to internals).

Reply:
This comment contradicts the comment at the top of the review.  Would it be preferable to inform the users via this documentation that zod effects are not supported, or that they can be supported by that requires overriding the `validateInput` function?

Reply:
Zod effects are basically some extra "application level" transformations that are not transferable to JSON Schema. But the framework (under the hoods) converts it to a JSON Schema. So to do something like "zod effects" you have  to specify `inputSchema` without them and do post-validation manually later in `validateSchema` method.


>The `inputSchema` definition must always be an object and pass validation by the `validateSchema()` function defined in [schema.ts](../src/internals/helpers/schema.ts).

This statement may sound like you are forcing the user to implement this feature. Either remove the statement or provide a simple example.